### PR TITLE
initial client side code for generating and fetching deviceEKs

### PR DESCRIPTION
**Pull request by oconnor663**
This hand waves all the details of key storage. It's just enough to pass
a unit test.

r? @joshblum

### CODE REVIEW

File: go/ephemeral/ephemeral.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		err = json.Unmarshal([]byte(element.MerklePayload), &signedMerkleRoot)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check whether the key is expired. This isn't considered an error,
+		// since the server doesn't do this check for us. We log these cases
+		// and skip them.
+		ageSecs := currentMerkleRoot.Ctime() - signedMerkleRoot.Ctime
+		if ageSecs > KeyLifetimeSecs {
```
Comment (joshblum) [inline go/ephemeral/ephemeral.go]: A key could be valid if it's older than a week if there was a gap in generation -- we should check this here right?

File: protocol/avdl/keybase1/ephemeral.avdl [hunk 2]
```diff
@@ -0,0 +1,14 @@
+@namespace("keybase.1")
+
+protocol ephemeral {
+  import idl "common.avdl";
+
+  record DeviceEkMetadata {
```
Comment (joshblum) [inline protocol/avdl/keybase1/ephemeral.avdl]: nit but let's call this `DeviceEKMetadata`

File: go/ephemeral/ephemeral.go (hunk 1 above)
Comment (joshblum) [reply go/ephemeral/ephemeral.go]: oops nvm this should only be checked for key deletion

File: go/ephemeral/ephemeral_test.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	require.NoError(t, err)
+
+	metadata, err := PublishNewDeviceEK(context.Background(), tc.G)
+	require.NoError(t, err)
+
+	fetchedDevices, err := GetOwnDeviceEKs(context.Background(), tc.G)
+	require.NoError(t, err)
+
+	require.Equal(t, 1, len(fetchedDevices))
+	require.Equal(t, metadata, fetchedDevices[0])
```
Comment (joshblum) [inline go/ephemeral/ephemeral_test.go]: let's verify the generation number, once we do the storage half we can verify beyond the hardcoded value !

File: go/ephemeral/ephemeral.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+const KeyLifetimeSecs = 60 * 60 * 24 * 7 // one week
+
+const EphemeralSeedSize = 32
```
Comment (joshblum) [inline go/ephemeral/ephemeral.go]: `EKSeedSize`

File: go/ephemeral/ephemeral.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+func makeNewRandomSeed() (seed [EphemeralSeedSize]byte, err error) {
+	bs, err := libkb.RandBytes(32)
+	if err != nil {
+		return seed, err
+	}
+	return libkb.MakeByte32(bs), nil
+}
+
+type DeviceEphemeralSeed [EphemeralSeedSize]byte
```
Comment (joshblum) [inline go/ephemeral/ephemeral.go]: `DeviceEKSeed`

File: protocol/avdl/keybase1/ephemeral.avdl (hunk 2 above)
Comment (oconnor663) [reply protocol/avdl/keybase1/ephemeral.avdl]: I tried that and some LintVetCheckerHookBotAutomatonFactory complained :(

**Comment by oconnor663** [comment]
Ready for a second look. Hopefully CircleCI will pass this time.

File: go/ephemeral/ephemeral.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+		// Decode the signed JSON.
+		var signedMetadata keybase1.DeviceEkMetadata
+		err = json.Unmarshal(payload, &signedMetadata)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check the hash of the Merkle payload, confirm it matches what was signed, and parse it.
+		computedHash := sha256.Sum256([]byte(element.MerklePayload))
```
Comment (joshblum) [inline go/ephemeral/ephemeral.go]: can you convert `element.MerklePayload` to `MerkleRootPayload` type and we can just call `ShortHash()` on that?

Comment (joshblum) [reply go/ephemeral/ephemeral.go]: nvm i see we want to make sure it's good before we parse

**Pull request merged**